src/Raven.Server/Documents/Indexes/QueryOperationContext.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Diagnostics;
+using Raven.Server.Documents.Queries;
+using Raven.Server.ServerWide;
+using Raven.Server.ServerWide.Context;
+
+namespace Raven.Server.Documents.Indexes
+{
+    public class QueryOperationContext : IDisposable
+    {
+        public readonly DocumentsOperationContext Documents;
+
+        public readonly TransactionOperationContext Server;
+
+        private IDisposable _releaseDocuments;
+        private IDisposable _releaseServer;
+
+        private QueryOperationContext(DocumentsOperationContext documentsContext, bool releaseDocumentsContext, bool needsServerContext)
+        {
+            if (documentsContext is null)
+                throw new ArgumentNullException(nameof(documentsContext));
+
+            Documents = documentsContext;
+
+            if (releaseDocumentsContext)
+                _releaseDocuments = Documents;
+
+            if (needsServerContext)
+                _releaseServer = documentsContext.DocumentDatabase.ServerStore.ContextPool.AllocateOperationContext(out Server);
+        }
+
+        private QueryOperationContext(DocumentDatabase database, bool needsServerContext)
+        {
+            if (database is null)
+                throw new ArgumentNullException(nameof(database));
+
+            _releaseDocuments = database.DocumentsStorage.ContextPool.AllocateOperationContext(out Documents);
+
+            if (needsServerContext)
+                _releaseServer = database.ServerStore.ContextPool.AllocateOperationContext(out Server);
+        }
+
+        private QueryOperationContext(DocumentDatabase database, Index index)
+        {
+            if (database is null)
+                throw new ArgumentNullException(nameof(database));
+            if (index is null)
+                throw new ArgumentNullException(nameof(index));
+
+            _releaseDocuments = database.DocumentsStorage.ContextPool.AllocateOperationContext(out Documents);
+
+            if (index.Definition.HasCompareExchange)
+                _releaseServer = database.ServerStore.ContextPool.AllocateOperationContext(out Server);
+        }
+
+        public IDisposable OpenReadTransaction()
+        {
+            var documentsTx = Documents.OpenReadTransaction();
+            RavenTransaction serverTx = null;
+            if (Server != null)
+                serverTx = Server.OpenReadTransaction();
+
+            return new DisposeTransactions(documentsTx, serverTx);
+        }
+
+        public void CloseTransaction()
+        {
+            Documents.CloseTransaction();
+            Server?.CloseTransaction();
+        }
+
+        public bool AreTransactionsOpened()
+        {
+            var opened = Documents.Transaction?.Disposed == false;
+
+            if (Server != null)
+                opened |= Server.Transaction?.Disposed == false;
```

Comment:
Is it allowed that one tx is open while another one is closed? Maybe let's add a check which will ensure that _both_ transactions are either open or closed?

Reply:
Done